### ISSUE TITLE
r.subrequest not working with post request from browser through http2 but curl works

### ISSUE THREAD

**Issue by abhayranjan7** [nginx/njs#338]
I am using njs v0.4.3. I have an Uri which accepts GET and POST methods. Now with curl both of these methods are working properly but from browser(Mozilla and Edge) only GET is working. Here is my nginx and njs configs:

**njs code:**

```js
function hello(r) {
        function response_cb(res) {
                r.return(res.status, res.responseBody);
        }
        r.subrequest('/api/foo', { method: r.method, body: r.requestBody }, response_cb);
}
export default {hello};
```

**nginx conf:**

```
location  = /api/foo {
    proxy_pass                                  http://upstream_server;
    proxy_pass_request_body                     on;
    proxy_set_header  Host                      $http_host;
    proxy_set_header  X-Real-IP                 $remote_addr;
    proxy_set_header  X-Forwarded-For           $proxy_add_x_forwarded_for;
    proxy_set_header  X-Forwarded-Proto         $scheme;
    proxy_read_timeout                          900;
}
```

I have checked the headers like Content-Type and Accept just to make sure that headers are fine. Also i tried to add debug logs inside callback function but it's not printing. So seems like POST request is not reaching till callback function, only GET request log is getting printed. No error logs are there which is making debugging difficult.
I have changed details like URI and other names from the production code just to make the explanation simple.

Hi @xeioex @drsm It would be really nice if you can check this issue. Thanks in advance.

**Comment by lexborisov** [comment]
Hi @abhayranjan7 

I am trying to reproduce your problem.
Below is my configuration:

**nginx.comf**
```nginx
user www;

error_log /dev/stdout info;

daemon  off;
master_process off;

events {
    worker_connections 1024;
}

http {
    default_type  text/html;
    access_log /dev/stdout;

    js_import module.js;
    js_import hello.js;

    server {
        listen       80;
        server_name  localhost;

        charset utf-8;

        location / {
            js_content hello.hello;
        }

        location = /api/foo {
            proxy_pass                                  http://localhost:8002;
            proxy_pass_request_body                     on;
            proxy_set_header  Host                      $http_host;
            proxy_set_header  X-Real-IP                 $remote_addr;
            proxy_set_header  X-Forwarded-For           $proxy_add_x_forwarded_for;
            proxy_set_header  X-Forwarded-Proto         $scheme;
            proxy_read_timeout                          900;
        }
    }

    server {
        listen       8002;
        server_name  localhost;

        charset utf-8;

        location = /api/foo {
            js_content module.hello;
        }
    }
}

```

**hello.js**
```JS
function hello(r) {
    function response_cb(res) {
        r.return(res.status, res.responseBody);
    }

    r.subrequest('/api/foo', { method: r.method, body: r.requestBody }, response_cb);
}

export default {hello};
```

**module.js**
```JS
function hello(r) {
    var form = "<form action='/' method='post'><button>POST</button></form>";

    r.return(200, form + "<br>" + njs.dump(r));
}

export default {hello};
```

In `Chrome Version 84.0.4147.135` method `GET` and `POST` work fine.

POST request result:

```JSON
   "method":"POST",
   "uri":"/api/foo",
   "headersIn":{
      "Host":"127.0.0.1",
      "X-Real-IP":"127.0.0.1",
      "X-Forwarded-For":"127.0.0.1",
      "X-Forwarded-Proto":"http",
      "Connection":"close",
      "Content-Length":"0",
      "Cache-Control":"max-age=0",
      "Upgrade-Insecure-Requests":"1",
      "User-Agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.135 Safari/537.36",
      "Origin":"http://127.0.0.1",
      "Content-Type":"application/x-www-form-urlencoded",
      "Accept":"text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
      "Sec-Fetch-Site":"same-origin",
      "Sec-Fetch-Mode":"navigate",
      "Sec-Fetch-User":"?1",
      "Sec-Fetch-Dest":"document",
      "Referer":"http://127.0.0.1/",
      "Accept-Encoding":"gzip, deflate, br",
      "Accept-Language":"ru-RU,ru;q=0.9,en-US;q=0.8,en;q=0.7"
   },
```

`nginx version: nginx/1.19.3`
`njs 0.4.4`

Maybe I misunderstood the example of your configuration?

**Comment by abhayranjan7** [comment]
Hi @lexborisov Thanks for the reply. I don't follow what module.js hello function is doing. But i tried again to reproduce my issue. Now i am seeing that, only GET request is working with http2 not POST. When i am switching to http1.1 my same code is working for both POST and GET from Mozilla. My ask is does njs support POST request with http2? Also your code is http1.1 as listen directive is not having http2.

**Comment by abhayranjan7** [comment]
I am seeing a mysterious case here. If request is reaching njs function **hello** in my case above with empty request body. r.subrequest is able to create POST create with request body. But it's failing to create subrequest if client request body is not empty . This is same for all methods. This is the new njs script file.
`function hello(r) {
    function response_cb(res) {
        r.return(res.status, res.responseBody);
    }
    var key = ("test");
    var key_base64 = key.toString('base64');
    var put_json = JSON.stringify({ key : key_base64 });
    r.subrequest('/api/foo', { method: 'POST', body:  put_json }, response_cb);
}

export default {hello};`

**Comment by xeioex** [comment]
@abhayranjan7 

Thanks for report, looks like you found a bug in nginx http2 code (not related to njs). Currently we are working on it.

**Comment by abhayranjan7** [comment]
Hi @xeioex thanks for the confirmation. Can you please tell the expected ETA for the fix(If you have any). It will help estimate my current work.

**Comment by xeioex** [comment]
@abhayranjan7 

it will be released with the upcoming nginx-1.19.3 release, by the beginning of October.

**Comment by abhayranjan7** [comment]
Thanks @xeioex for the update.

**Comment by xeioex** [comment]
@abhayranjan7 BTW, here is the patch for nginx. Feel free to test it.

```diff
# HG changeset patch
# User Maxim Dounin <mdounin@mdounin.ru>
# Date 1600356599 -10800
#      Thu Sep 17 18:29:59 2020 +0300
# Node ID 6441ade61932a0005610d8720a554e711be92acc
# Parent  052ecc68d35038b1b4adde12efe6249a92055f09
HTTP/2: run posted requests after reading body.

HTTP/2 code failed to run posted requests after calling the request body
handler, and this resulted in connection hang if a subrequest was created
in the body handler and no other actions were made.

diff --git a/src/http/v2/ngx_http_v2.c b/src/http/v2/ngx_http_v2.c
--- a/src/http/v2/ngx_http_v2.c
+++ b/src/http/v2/ngx_http_v2.c
@@ -1058,6 +1058,7 @@ ngx_http_v2_state_read_data(ngx_http_v2_
     size_t                   size;
     ngx_buf_t               *buf;
     ngx_int_t                rc;
+    ngx_connection_t        *fc;
     ngx_http_request_t      *r;
     ngx_http_v2_stream_t    *stream;
     ngx_http_v2_srv_conf_t  *h2scf;
@@ -1076,6 +1077,7 @@ ngx_http_v2_state_read_data(ngx_http_v2_
     }
 
     r = stream->request;
+    fc = r->connection;
 
     if (r->reading_body && !r->request_body_no_buffering) {
         ngx_log_debug0(NGX_LOG_DEBUG_HTTP, h2c->connection->log, 0,
@@ -1101,6 +1103,8 @@ ngx_http_v2_state_read_data(ngx_http_v2_
             ngx_http_finalize_request(r, rc);
         }
 
+        ngx_http_run_posted_requests(fc);
+
     } else if (size) {
         buf = stream->preread;
```

**Comment by xeioex** [comment]
@abhayranjan7 fix is releases with nginx-1.19.3.

**Comment by abhayranjan7** [comment]
Thanks for the update @xeioex
